### ISSUE TITLE
Zero Intelligence API: Freemium RAG Data & Rules Distribution

### ISSUE THREAD

**Issue by curphey** [crashappsec/zero#108]
## Summary

Create a freemium intelligence distribution model for Zero's RAG data, detection rules, and trained models. Similar to antivirus signature distribution:

- **Free/OSS Tier**: Bundled rules (updated every 3-6 months), no training data contribution
- **Registered API Tier**: Fresh rules via API key, rate-limited, users contribute validation feedback

## Background

Zero's intelligence layer consists of:
- RAG patterns (`/rag/` directory) for technology detection
- Semgrep rules (synced via `zero feeds semgrep`)
- Detection patterns in analyzers
- Confidence scoring models trained on feedback

This intelligence is valuable and should be:
1. Available for free/OSS users (stale data)
2. Fresh for registered users who contribute back
3. Protected from freeloading (rate limited)

## Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                    Zero Intelligence Service                     │
├─────────────────────────────────────────────────────────────────┤
│                                                                  │
│  ┌──────────────┐    ┌──────────────┐    ┌──────────────────┐   │
│  │   RAG Data   │    │  Semgrep     │    │  Trained Models  │   │
│  │  (Patterns)  │    │   Rules      │    │  (Confidence)    │   │
│  └──────────────┘    └──────────────┘    └──────────────────┘   │
│           │                  │                    │              │
│           └──────────────────┼────────────────────┘              │
│                              ▼                                   │
│                    ┌─────────────────┐                          │
│                    │ Intelligence API │                          │
│                    │  (Rate Limited)  │                          │
│                    └─────────────────┘                          │
│                              │                                   │
│         ┌────────────────────┼────────────────────┐              │
│         ▼                    ▼                    ▼              │
│  ┌─────────────┐    ┌──────────────┐    ┌──────────────┐        │
│  │ Free Tier   │    │ Registered   │    │  Enterprise  │        │
│  │ (Bundled)   │    │   (API Key)  │    │  (Premium)   │        │
│  └─────────────┘    └──────────────┘    └──────────────┘        │
│                              │                                   │
│                              ▼                                   │
│                    ┌─────────────────┐                          │
│                    │ Feedback Loop   │                          │
│                    │ (Training Data) │                          │
│                    └─────────────────┘                          │
└─────────────────────────────────────────────────────────────────┘
```

### Free/OSS Tier (Bundled)
- **Rules**: Bundled with release (updated every 3-6 months)
- **RAG Data**: Static, bundled version
- **Updates**: Major releases only
- **Training**: No contribution required
- **Support**: Community only

### Registered API Tier (Free with API Key)
- **Rules**: Live API access to latest rules
- **RAG Data**: Fresh patterns (updated weekly)
- **Updates**: Real-time via API
- **Training**: **Required** - must submit validation feedback
- **Rate Limits**: 1000 requests/day
- **Support**: Email support

### Enterprise Tier (Paid)
- **Rules**: Priority access, custom rules
- **RAG Data**: Full dataset + custom patterns
- **Updates**: Real-time + early access
- **Training**: Optional (higher rate limits if contributing)
- **Rate Limits**: Unlimited
- **Support**: Dedicated support

### Registration & Auth
```
POST /api/v1/register
  { email, org_name, usage_type: "oss" | "commercial" }
  → { api_key, tier, rate_limit }

POST /api/v1/auth/verify
  Header: X-API-Key: <key>
  → { valid, tier, remaining_requests, usage_stats }
```

### Intelligence API
```
GET /api/v1/intelligence/rag/{category}
  Header: X-API-Key: <key>
  → { patterns: [], version, last_updated }
  Categories: technology, security, quality, devops

GET /api/v1/intelligence/rules/{type}
  Header: X-API-Key: <key>
  → { rules: [], version, checksum }
  Types: semgrep, secrets, crypto

GET /api/v1/intelligence/models/{name}
  Header: X-API-Key: <key>
  → { model_data, version, accuracy_metrics }
  Models: confidence-tech-id, confidence-secrets
```

### Feedback Submission (Required for Registered Tier)
```
POST /api/v1/feedback/validation
  Header: X-API-Key: <key>
  {
    finding_type: "technology" | "secret" | "vulnerability",
    finding_hash: "<hash>",
    validation: "true_positive" | "false_positive" | "false_negative",
    context: { file, line, content_sample },
    metadata: { analyzer_version, confidence_score }
  }
  → { accepted: true, contribution_credit: 10 }

GET /api/v1/feedback/stats
  Header: X-API-Key: <key>
  → { total_contributions, accuracy_improvement, tier_benefits }
```

### Rate Limiting
```
GET /api/v1/rate-limit/status
  Header: X-API-Key: <key>
  → { 
    tier: "registered",
    limit: 1000,
    remaining: 847,
    reset_at: "2024-01-14T00:00:00Z",
    contribution_bonus: 100  // Extra requests for feedback
  }
```

## Feedback → Training Loop

```
User Validation                     Training Pipeline
      │                                    │
      ▼                                    │
┌─────────────┐                           │
│ Submit FP   │ ──────────────────────────┤
│ via API     │                           │
└─────────────┘                           │
      │                                    │
      ▼                                    ▼
┌─────────────┐                    ┌─────────────┐
│ Store in    │                    │ Aggregate   │
│ Feedback DB │ ────────────────►  │ Feedback    │
└─────────────┘                    └─────────────┘
                                          │
                                          ▼
                                   ┌─────────────┐
                                   │ Retrain     │
                                   │ Models      │
                                   └─────────────┘
                                          │
                                          ▼
                                   ┌─────────────┐
                                   │ Update RAG  │
                                   │ Patterns    │
                                   └─────────────┘
                                          │
                                          ▼
                                   ┌─────────────┐
                                   │ Publish to  │
                                   │ API         │
                                   └─────────────┘
```

## CLI Integration

```bash
# Register for API access
zero auth register --email user@example.com --org "My Org"

# Check API status
zero auth status
# → Tier: Registered
# → API Key: zk_xxx...xxx
# → Requests remaining: 847/1000
# → Contributions: 42 (earning +100 bonus requests)

# Sync fresh intelligence (requires API key)
zero feeds sync --live
# → Syncing RAG patterns... done (127 updated)
# → Syncing Semgrep rules... done (45 new)
# → Syncing confidence models... done

# Submit feedback (increases rate limit)
zero feedback submit --type false-positive --finding-id abc123
# → Feedback submitted. +10 contribution credit.

# Use bundled (offline) data
zero feeds sync --offline
# → Using bundled data (v2024.01, 3 months old)
```

### Phase 1: API Infrastructure
- [ ] Create intelligence API service (separate repo or microservice)
- [ ] Implement API key generation and storage
- [ ] Add rate limiting middleware
- [ ] Create registration flow

### Phase 2: Intelligence Packaging
- [ ] Package RAG data for API distribution
- [ ] Create rule versioning system
- [ ] Implement differential updates (only changed patterns)
- [ ] Bundle static data with releases

### Phase 3: Feedback Collection
- [ ] Extend existing feedback system for API submission
- [ ] Create feedback aggregation pipeline
- [ ] Implement contribution tracking
- [ ] Add rate limit bonuses for contributors

### Phase 4: Training Pipeline
- [ ] Create automated model retraining from feedback
- [ ] Implement A/B testing for rule changes
- [ ] Add accuracy metrics dashboard
- [ ] Set up CI/CD for intelligence updates

### Phase 5: CLI Integration
- [ ] Add `zero auth register` command
- [ ] Modify `zero feeds` to support API vs bundled
- [ ] Add contribution tracking to CLI
- [ ] Create offline fallback mode

## Related Issues

- #91 - Finding Validation UI (user-facing validation)
- #45 - Finding Validation System (backend feedback storage)
- #40 - Supabase storage backend (potential for cloud feedback storage)

## Security Considerations

- API keys should be revocable
- Rate limiting must prevent abuse
- Feedback should be validated to prevent poisoning
- No PII should be collected in feedback (content samples sanitized)
